### PR TITLE
erts: Implement AF_VSOCK domain in prim_socket

### PR DESCRIPTION
**Pull request by meeq**
Context
=======

VSock is a mechanism for socket communication between virtual machines and their host operating system. The host and each VM are assigned a 32-bit CID (Context IDentifier) and may connect or bind to a 32-bit port number.

The Linux kernel provides support for `AF_VSOCK` sockets in the [`linux/vm_sockets.h` header](https://github.com/torvalds/linux/blob/master/include/uapi/linux/vm_sockets.h) and is documented in the [`vsock(7)` manual page](https://man7.org/linux/man-pages/man7/vsock.7.html).

[AWS EC2 Nitro Enclaves rely on `AF_VSOCK` as the sole communication channel between the parent instance and its secure enclaves.](https://docs.aws.amazon.com/enclaves/latest/user/nitro-enclave-concepts.html#term-socket)

Implementation
==============

My approach was to introduce a new `vsock` Domain in the `socket` module, which is supported by the `prim_socket` NIF. The `vsock` Domain exposes the `port` and `cid` fields of the `sockaddr_vm` struct. Otherwise, this PR generally follows the existing patterns established by the other domains.

Known limitations
==================

`AF_VSOCK` is only supported on Linux at this time:

  * VMware (VMCI) has been supported since Linux 3.9
  * KVM (virtio) has been supported since Linux 4.8
  * Hyper-V has been supported since Linux 4.14
  * Local communication using `VMADDR_CID_LOCAL` has been supported since Linux 5.6

XNU kernel provides low-level `AF_VSOCK` support in [`sys/vsock.h`](https://opensource.apple.com/source/xnu/xnu-7195.60.75/bsd/sys/vsock.h.auto.html), however macOS requires the use of a higher-level [Virtualization framework](https://developer.apple.com/documentation/virtualization) that manages [`VZVirtioSocketDevice` objects](https://developer.apple.com/documentation/virtualization/vzvirtiosocketdevice) to work with VSock connections. **This PR makes no attempt to interface with macOS's Virtualization framework.**

This PR does not currently include any tests. It is not clear to me how best to write tests for this functionality, but I welcome any suggestions.

### CODE REVIEW

**Comment by bmk** [comment]
Looks like you have updated erts/configure directly, which is not the way to do it
(I am writing from memory, so it may not be *exactly* right):
`
1) Update erts/configure.ac
2) Run ./otp_build update_configure --no-commit
3) Add updated erts/configure to repo: git add erts/configure
4) [possibly] Clear out any other stuff changed by 'update_configure': git checkout ...
5) git commit
`

**Comment by meeq** [comment]
@bmk  I have incorporated the feature detection into `erts/configure.ac` and regenerated the `erts/configure` script. Your steps were accurate. Thank you for your feedback and instruction.

**Comment by RaimoNiskanen** [comment]
Now I got around to look at this, thank you for your contribution!

I am testing in the shell:
```erlang
1> {ok,L} = socket:open(vsock, stream).
{ok,{'$socket',#Ref<0.1855165874.552468482.132081>}}
2> socket:bind(L, #{family => vsock, cid = any}).
* 1:44: syntax error before: '}'
2> socket:bind(L, #{family => vsock, cid => any}).
ok
3> socket:listen(L).
ok
4> {ok,C} = socket:open(vsock, stream).
{ok,{'$socket',#Ref<0.1855165874.552468482.132111>}}
5> socket:sockname(L).
{ok,#{port => 206641836,family => vsock,cid => 4294967295}}
6> socket:connect(C, #{port => 206641836,family => vsock,cid => local}).
ok
7> socket:sockname(L).
{ok,#{port => 206641836,family => vsock,cid => 4294967295}}
8> socket:sockname(C).
{ok,#{port => 206641837,family => vsock,cid => 4294967295}}
9> socket:peername(C).
{ok,#{port => 206641836,family => vsock,cid => 1}}
10> {ok,S} = socket:accept(L, 3000).
{ok,{'$socket',#Ref<0.1855165874.552468482.132150>}}
11> socket:sockname(S).
{ok,#{port => 206641836,family => vsock,cid => 1}}
12> socket:peername(S).
{ok,#{port => 206641837,family => vsock,cid => 1}}
13> socket:send(C, "hello world").
ok
14> socket:recv(S).
{ok,<<"hello world">>}
```

This should definitely be possible to turn into a smoke test case...

Some questions:
* I saw in the code that `any` should be shown as `-1`, but that does not seem to be the case here.  Did I misread?
* I guess that the idea is that an address (CID) should always be shown numerically, just like an IP address, but aliases such as `any`, `host`, ...  can be set.  Is that correct?
* The address family should also be implemented in `socket:supports/*` and `socket:is_supported/*`.  Compare with AF_LOCAL.
* Then, test cases can be written to check for AF_VSOCK support and skip on all platforms where it is unsupported. See lib/kernel/test/socket_SUITE.erl the `api_basic_cases/0` test cases named `*_tcpL`, `*udpL`, `*_dgram_local`, and `*_stream_local`. These should be possible to augment for AF_VSOCK.

**Comment by meeq** [comment]
Regarding your comments @RaimoNiskanen :

> I saw in the code that `any` should be shown as `-1`, but that does not seem to be the case here.  Did I misread?

`any` is only shown as `-1` when the sockaddr is formatted. This is consistent with how the `inet` family stores `any` internally as `{0,0,0,0}` but special-cases it for formatting.

> I guess that the idea is that an address (CID) should always be shown numerically, just like an IP address, but aliases such as `any`, `host`, ...  can be set.  Is that correct?

Yes, just like the IP address aliases for `'any' | 'broadcast' | 'loopback'` get converted to a 4-tuple by `esock_decode_in_addr`, the vsock family supports aliases for `'any' | 'local' | 'host' | 'hypervisor'` that resolve in `esock_decode_sockaddr_vm`

>  The address family should also be implemented in `socket:supports/*` and `socket:is_supported/*`.  Compare with AF_LOCAL.

Thanks for this feedback. I will incorporate this.

> Then, test cases can be written to check for AF_VSOCK support and skip on all platforms where it is unsupported. See lib/kernel/test/socket_SUITE.erl the `api_basic_cases/0` test cases named `*_tcpL`, `*udpL`, `*_dgram_local`, and `*_stream_local`. These should be possible to augment for AF_VSOCK.

Thank you for pointing me in the right direction here. I will implement this.

**Comment by RaimoNiskanen** [comment]
@meeq wrote:
> This is consistent with how the inet family stores any internally as {0,0,0,0} but special-cases it for formatting.

Right.  Great!

> I will implement this.

Excellent!

**Comment by IngelaAndin** [comment]
@meeq ping

**Comment by meeq** [comment]
I will try to wrap up the outstanding feedback on this within the next week.

**Comment by IngelaAndin** [comment]
@meeq  There are now some failing checks for this PR, I am unsure if they might have happened along the way (where not there when you first updated the PR). Could you have a look.

**Comment by RaimoNiskanen** [comment]
As I can see test cases are still missing.

Also, it is getting hard to follow what changes this branch introduces since you merge in 'master' now and then.  We would prefer that you rebase the whole branch onto a 'master' instead, to keep the topic branch as a straight sequence of commits.

**Comment by IngelaAndin** [comment]
@meeq are you planning to continue this? We do not have a own interest in this PR, at least not at moment. So either you need to work on making it mergeable or we will close it due to inactivity.

**Comment by meeq** [comment]
Hello, thank you for the reminder. I have not had time to work on this branch due to other commitments. Let's close this for now, and I will resubmit at a future date when I have addressed all of the outstanding feedback.

**Comment by IngelaAndin** [comment]
@meeq  ok thank you and looking forward to your future contribution.
